pkg/ptrace/ptrace_linux.go

```diff
@@ -0,0 +1,410 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//go:build cgo
+
+package ptrace
+
+/*
+#include <stdint.h>
+struct iovec {
+	intptr_t iov_base;
+	size_t iov_len;
+};
+*/
+import "C"
+
+import (
+	"bytes"
+	"debug/elf"
+	"fmt"
+	"io/ioutil"
+	"strconv"
+	"strings"
+	"syscall"
+	"unsafe"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/mapreader"
+)
+
+const waitPidErrorMessage = "waitpid ret value: %d"
+
+// If it's on 64-bit platform, `^uintptr(0)` will get a 64-bit number full of one.
+// After shifting right for 63-bit, only 1 will be left. Than we got 8 here.
+// If it's on 32-bit platform, After shifting nothing will be left. Than we got 4 here.
+const ptrSize = 4 << uintptr(^uintptr(0)>>63)
+
+var threadRetryLimit = 10
+
+// TracedProgram is a program traced by ptrace
+type TracedProgram struct {
+	pid     int
+	tids    []int
+	Entries []mapreader.Entry
+
+	backupRegs *syscall.PtraceRegs
+	backupCode []byte
+
+	logger logr.Logger
+}
+
+// Pid return the pid of traced program
+func (p *TracedProgram) Pid() int {
+	return p.pid
+}
+
+func waitPid(pid int) error {
+	ret := waitpid(pid)
+	if ret == pid {
+		return nil
+	}
+
+	return errors.Errorf(waitPidErrorMessage, ret)
+}
+
+// Trace ptrace all threads of a process
+func Trace(pid int, logger logr.Logger) (*TracedProgram, error) {
+	traceSuccess := false
+
+	tidMap := make(map[int]bool)
+	retryCount := make(map[int]int)
+	for {
```

Comment:
I do know these code works well for a long time, but hardly can I understand what for loop here does.

Reply:
I suggest to add some comment here.

Reply:
Added